File: app/src/main/java/info/nightscout/androidaps/setupwizard/SWEventListener.java
package info.nightscout.androidaps.setupwizard;

import android.content.Context;
import android.widget.LinearLayout;
import android.widget.TextView;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import info.nightscout.androidaps.MainApp;
import info.nightscout.androidaps.events.EventStatus;
import info.nightscout.androidaps.plugins.bus.RxBus;
import info.nightscout.androidaps.setupwizard.elements.SWItem;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;


public class SWEventListener extends SWItem {
    private static Logger log = LoggerFactory.getLogger(SWEventListener.class);
    private CompositeDisposable disposable = new CompositeDisposable();

    private int textLabel = 0;
    private String status = "";
    TextView textView;
    SWDefinition definition;

    SWEventListener(SWDefinition definition, Class clazz) {
        super(Type.LISTENER);
        this.definition = definition;
        disposable.add(RxBus.INSTANCE
                .toObservable(clazz)
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(event -> {
                    status = ((EventStatus) event).getStatus();
                    if (textView != null)
                        textView.setText((textLabel != 0 ? MainApp.gs(textLabel) : "") + " " + status);
                })
        );

    }

    public SWEventListener label(int newLabel) {
        this.textLabel = newLabel;
        return this;
    }

    SWEventListener initialStatus(String status) {
        this.status = status;
        return this;
    }

    @Override
    public void generateDialog(LinearLayout layout) {
        Context context = layout.getContext();

        textView = new TextView(context);
        textView.setId(layout.generateViewId());
        textView.setText((textLabel != 0 ? MainApp.gs(textLabel) : "") + " " + status);
        layout.addView(textView);
    }
}

File: app/src/main/java/info/nightscout/androidaps/setupwizard/SWEventListener.kt
package info.nightscout.androidaps.setupwizard

import android.annotation.SuppressLint
import android.view.View
import android.widget.LinearLayout
import android.widget.TextView
import dagger.android.HasAndroidInjector
import info.nightscout.androidaps.events.EventStatus
import info.nightscout.androidaps.setupwizard.elements.SWItem
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.CompositeDisposable

class SWEventListener constructor(
    injector: HasAndroidInjector,
    clazz: Class<out EventStatus>
) : SWItem(injector, Type.LISTENER) {

    private val disposable = CompositeDisposable()
    private var textLabel = 0
    private var status = ""
    private var textView: TextView? = null
    private var visibilityValidator: SWValidator? = null

    // TODO: Adrian how to clear disposable in this case?
    init {
        disposable.add(rxBus
            .toObservable(clazz)
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe { event: Any ->
                status = (event as EventStatus).getStatus(resourceHelper)
                @SuppressLint("SetTextI18n")
                textView?.text = (if (textLabel != 0) resourceHelper.gs(textLabel) else "") + " " + status
            }
        )
    }

    override fun label(label: Int): SWEventListener {
        textLabel = label
        return this
    }

    fun initialStatus(status: String): SWEventListener {
        this.status = status
        return this
    }

    fun visibility(visibilityValidator: SWValidator): SWEventListener {
        this.visibilityValidator = visibilityValidator
        return this
    }

    @SuppressLint("SetTextI18n")
    override fun generateDialog(layout: LinearLayout) {
        val context = layout.context
        textView = TextView(context)
        textView?.id = View.generateViewId()
        textView?.text = (if (textLabel != 0) resourceHelper.gs(textLabel) else "") + " " + status
        layout.addView(textView)
    }

    override fun processVisibility() {
        if (visibilityValidator != null && !visibilityValidator!!.isValid) textView?.visibility = View.GONE else textView?.visibility = View.VISIBLE
    }
}